### ISSUE TITLE
wc_add_order_item_meta is not applying to customer order received emails in woocommerce3.x

### ISSUE THREAD

**Issue by sideburn** [woocommerce/woocommerce#15285]
### EXPLANATION OF THE ISSUE

In Woocommerce 3.x, when inserting metadata on a completed order, the metadata no longer appears in the Order Received emails. It does appear on the Order Received page though.

### STEPS TO REPRODUCE THE ISSUE

Create a plugin that adds a woocommerce_order_status_completed action and have that function make a call to wc_add_order_item_meta. See below:

` add_action('woocommerce_order_status_completed', array($this, 'action_woocommerce_order_status_completed'));

public function action_woocommerce_order_status_completed($order_id) {
        wc_add_order_item_meta($item_id, __('Serial Number','wlmg'), "ABC-123-XYZ");
}`

This will get called before email-order-items.php but the metadata will no longer be in the array.
next order-details-item.php will get called and the metadata will now be there.

### SYSTEM STATUS

<details>
```
`
### WordPress Environment ###

Home URL: http://www.woocommerce.dev
Site URL: http://www.woocommerce.dev
WC Version: 3.0.7
Log Directory Writable: ✔
WP Version: 4.7.5
WP Multisite: –
WP Memory Limit: 2 GB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: Apache/2.4.12 (Unix) OpenSSL/1.0.1m PHP/5.5.24 mod_perl/2.0.8-dev Perl/v5.16.3
PHP Version: ❌ 5.5.24 - We recommend a minimum PHP version of 5.6. See: How to update your PHP version
PHP Post Max Size: 10 GB
PHP Time Limit: 0
PHP Max Input Vars: 1000
cURL Version: 7.30.0
OpenSSL/1.0.1m

SUHOSIN Installed: –
MySQL Version: 5.6.24
Max Upload Size: 10 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.0.7
WC Database Prefix: wp_
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔
MaxMind GeoIP Database: ❌ The MaxMind GeoIP Database does not exist - Geolocation will not function. You can download and install it manually from http://dev.maxmind.com/geoip/legacy/geolite/ to the path: . Scroll down to "Downloads" and download the "Binary / gzip" file next to "GeoLite Country". Please remember to uncompress GeoIP.dat.gz and upload the GeoIP.dat file only.

### Security ###

Secure connection (HTTPS): ❌Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (2) ###

WooCommerce Metadata Test: by Tavis Hord – 1.0
WooCommerce: by Automattic – 3.0.7

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #4 - /shop/
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My account: #7 - /my-account/

### Theme ###

Name: dark-child
Version: 1.0.0
Author URL: http://www.sideburn.com
Child Theme: ✔
Parent Theme Name: Dark
Parent Theme Version: 1.1.7
Parent Theme Author URL: http://minathemes.com
WooCommerce Support: ✔

### Templates ###

Archive Template: Your theme has a woocommerce.php file
you will not be able to override the woocommerce/archive-product.php custom template since woocommerce.php has priority over archive-product.php. This is intended to prevent display issues.

Overrides: –
`
```
</details>

**Comment by mikejolley** [comment]
Without reloading the object it won't see this update since it's not added via CRUD. 

woocommerce_order_status_completed gives you the ID and order object.

Use this method on the order object:

https://github.com/woocommerce/woocommerce/blob/master/includes/abstracts/abstract-wc-data.php#L345

That should ensure the order knows your new field exists.

**Comment by sideburn** [comment]
I have tried your suggestion but have not had any success yet.

Ive tried the following test as well and $exists is always false:

wc_add_order_item_meta($key, __('Hello World','wlmg'), "I am here");
$OrderItemObject = $items[$key];
$exists = $OrderItemObject->meta_exists('Hello World');

I have also tried calling meta_exists() on the order object itself with no luck. But I didn't expect that to work since I am adding order item metadata.

**Comment by mikejolley** [comment]
In that example, hello world is the value, not the key. meta_exists tests for keys.

**Comment by sideburn** [comment]
Here is some more descriptive code. I removed some unrelated routines inside the function to reduce it (I can email you the entire.php file is needed):

add_action('woocommerce_order_status_completed', array($this, 'action_woocommerce_order_status_completed'));

public function action_woocommerce_order_status_completed($order_id) {
	$order = new WC_Order($order_id);
	$items = $order->get_items();
	foreach ($items as $key => $value) {
		quantity = $value['quantity']
		$productID = $value['product_id'];
		$this->action_woocommerce_add_order_item_meta($key, array('quantity' => $quantity, 'product_id' => $productID));		
		
		//$exists is returning false and when I look at the meta_data in the debugger "'Serial Number(s)'" is still not there...
		$OrderItemObject = $items[$key];
		$exists = $OrderItemObject->meta_exists('Serial Number(s)');
	}					
}

public function action_woocommerce_add_order_item_meta($item_id, $values) {
	global $wpdb,$post;
	$licenseString = "123456";
	wc_add_order_item_meta($item_id, __('Serial Number(s)','wlmg'), $licenseString);
			
	$meta = wc_get_order_item_meta($item_id,null, true);
	//verified that $meta DOES contain my newly added data, but when email-order-items.php calls wc_display_item_meta( $item ),
	//the meta data is not there. When wc_display_item_meta( $item ) gets called a 2nd time and the "order details" page is displayed, the metadata IS there.
}

**Comment by sideburn** [comment]
I thought key would be 'Hello World’ and value is "I am here”.

I posted some full (but reduced) code in the thread. If you want the complete un-edited code or the php file, I can send that to you as well.
But the way to reproduce the issue is to simply call wc_add_order_item_meta() inside of a woocommerce_order_status_completed action and the email notifications will NOT contain the added order item meta but the order details page WILL contain them...

**Comment by mikejolley** [comment]
If I'm following that correct, you set `$items`, then add meta in a separate function (action_woocommerce_add_order_item_meta), then look at $items again. But $items will be stale and won't contain the change.

You'd need to load the items again to see that meta. Or you could interact with the item within your order object instead of using wc_add_order_item_meta.

**Comment by sideburn** [comment]
It doesn’t matter if it’s a separate function or not. I could make a simpler example that demonstrates the issue. Like I mentioned, the problem is simply that if you add item meta data inside of an order completed action, when Woocommerce sends the order completed notification email, the metadata that you added does not appear in the email.  But then when the order details page renders in the browser the meta data is there. I am
Adding a product serial number to the order item meta data and would like it to be included in the customers order completed notification email. If I add order item metadata. It’s some sort of timing issue. It used to work fine and then it broke in Woocommerce 3.x

**Comment by sideburn** [comment]
Actually my ver first example shows how to reproduce the issue:

` add_action('woocommerce_order_status_completed', array($this, 'action_woocommerce_order_status_completed'));

public function action_woocommerce_order_status_completed($order_id) {
wc_add_order_item_meta($item_id, __('Serial Number','wlmg'), "ABC-123-XYZ");
}`

With the above code, once you place an order and complete it, the email notifications will not contain the "serial number" metadata but the order details webpage WILL contain the meta data.

This is the issue I want to sort out how to solve.

**Comment by sideburn** [comment]
Here is a simpler example that can reproduce the issue (see comments in code):

add_action('woocommerce_order_status_completed', array($this, 'action_woocommerce_order_status_completed'));

public function action_woocommerce_order_status_completed($order_id) {
	$order = new WC_Order($order_id);
	$items = $order->get_items();
	foreach ($items as $key => $value) {
		wc_add_order_item_meta($key, __('Serial Number','wlmg'), "123456");

		//Read back the data to make sure it is there
		$meta = wc_get_order_item_meta($key,null, true);  //$meta DOES contain my newly added 'Serial Number' but the customer email notification does not display it. Why not?
	}					
}

**Comment by mikejolley** [comment]
Hmm looking at the email classes, some of the emails are actually sent during `woocommerce_order_status_completed` - have you attempted using an earlier priority?

```
add_action('woocommerce_order_status_completed', array($this, 'action_woocommerce_order_status_completed', 1 ) );
```

The customer completed email sends at priority 10.

**Comment by sideburn** [comment]
I have. I have also tried various other actions instead of the woocommerce_order_status_completed. 

I have tried:

woocommerce_order_status_pending_to_completed, woocommerce_order_status_pending_to_completed_notification, And various others... 

I also have set breakpoints in the debugger and ensured that the actions were getting called prior to sending the emails. I tried the hooks that are in the email classes as well. The ones with comments noting to let others add item metadata (I can’t recall off the top my head what those were called) but no luck. The odd thing is that immediately after the emails notifications send, the order details page appears and it does contain my metadata.

**Comment by mikejolley** [comment]
Just tested this code:

```
add_action('woocommerce_order_status_completed', 'action_woocommerce_order_status_completed', 5);

function action_woocommerce_order_status_completed($order_id) {
	$order = new WC_Order($order_id);
	$items = $order->get_items();

	foreach ($items as $key => $value) {
		wc_add_order_item_meta($key, __('Serial Number','wlmg'), "123456");

		//Read back the data to make sure it is there
		$meta = wc_get_order_item_meta($key,null, true);  //$meta DOES contain my newly added 'Serial Number' but the customer email notification does not display it. Why not?
	}
}
```

it appeared in my email.

**Comment by sideburn** [comment]
Hmm, I will try again with the exact code above. If it still fails I'll try a bare bones woocommerce and see if theres a plugin conflict somewhere.

**Comment by sideburn** [comment]
OK I have tried again with the exact code you used and still no luck. I disabled my theme and switched to the generic "Twenty Fifteen" theme. I disabled All plugins except for Woocommerce. I set a breakpoint in my code to ensure that the action was getting called and the metadata was being added. No metadata in the email but the meta data is there in the order details page... Not sure what our differences could be.

My Woocommerce version is 3.5.3
Wordpress version: 5.1

**Comment by mikejolley** [comment]
Are we looking at the same emails? I marked the order completed in the backend and looked at the customer facing email - thats where the meta was shown.

**Comment by sideburn** [comment]
Yes the customer order completed email. 
I am using a purchase gateway set in test mode. I then add the product to the cart and run the purchase through, the action gets called, the order item meta gets added, the customer email gets sent, and the webpage loads the order details page. 

The results I see are the customer email does not contain the order item metadata but the order details webpage does.

**Comment by mikejolley** [comment]
Try this;

- code above
- create order using cheque gateway. it will be on hold.
- mark order complete in admin.
- see email

**Comment by mikejolley** [comment]
If you view the source of the PP gateway, see which statuses it sets your order via the code. 

Also, consider where your serial code is running - is it available when that status change occurs.

**Comment by sideburn** [comment]
We only want to issue the serial number once the order is marked as completed. The odd thing is that I can insert meta data before the completed action. I’ve tried several other actions like pending completion, order status changed, etc. nothing seems to work unless I insert the meta data prior to checkout. But the order details page always displays the meta data.

**Comment by mikejolley** [comment]
Thats fine, but how does the test gateway you're using set the order status after checkout? I'm assuming it marks complete but what does that code look like. Something must be different there?

**Comment by sideburn** [comment]
I will look today but ANY gateway has the same issue. As long as the order completes. The gateways that I am using are Woocommerce Authorize.net CIM gateway, PayPal Checkout gateway, and PayPal gateway. All three methods have the same issue.

**Comment by sideburn** [comment]
I checked the Woocommerce Authorize.net CIM gateway and it is calling $this->set_status( apply_filters( 'woocommerce_payment_complete_order_status', $this->needs_processing() ? 'processing' : 'completed', $this->get_id(), $this ) );"

Then we fall into my action: action_woocommerce_order_status_completed

Then the email gets sent (missing my metadata)

and then then gateway calls

do_action( 'woocommerce_payment_complete', $this->get_id() );

**Comment by sideburn** [comment]
Any ideas on a fix? Can this issue be reopened since it appears to be a bug in Woocommerce?

**Comment by mikejolley** [comment]
I don't think I have access to that CIM gateway to see what `set_status` does. I suggest modifying the cheque gateway to set completed whilst testing this, and reproducing with that since it doesn't take a payment.

**Comment by sideburn** [comment]
Can you try PayPal or PayPal Express In sandbox mode? Those also fail. The only oddball is the Cheque gateway. Any gateway that actually competes an order will reproduce the issue.

I am able to reproduce the issue in sandbox / test mode on all these gateways. The problem is getting you or someone else to reproduce the issue in order to help me find a fix or apply a fix to Woocommerce.

**Comment by mikejolley** [comment]
Can you explain this? 

> The only oddball is the Cheque gateway.

Are you saying if completed is set **right away** there is no bug. If using a gateway which later sets completed status *the problem occurs*?

**Comment by sideburn** [comment]
Yes, that appears to be the case. 
When I force it to complete and use the Cheque gateway there is no bug. And like I mentioned before the issue cropped up in a woocommerce 3 update. Our site uses 3 gateways (authorize.net, PayPal, and PayPal express checkout). All three produce the problem on every transaction. We get several transactions a day selling software that issues a serial number to the customer. The serial number is normally in their order emails. Then it broke after updating woocommerce. It’s a pain to deal with because we need to as the customer to log back into the website and go to their account page to find their serial number.

**Comment by mikejolley** [comment]
Ok, so following the code, it's due to the way meta is added.

1. PayPal (or any other gateway) gets the order and updates it's status
2. The updated order fires off a hook to say "my status changed". It passes it's ID and itself to the action hook.
3. You code adds meta to an item. The item referenced by the order won't see this change - it already loaded the items. It won't see this change until it reloads items, or itself is reloaded.
4. The email template does not reload the order, it uses the passed order object.

So meta exists, but meta is not visible within the context of the email being sent.

To work around this:

1. When using `woocommerce_order_status_completed` hook, accept 2 args. First is ID, second is order object. Get the items from the order object rather than call `wc_get_order`.
2. Replace `wc_add_order_item_meta` with `$item->add_meta_data( $key, $value ); $item->save()`

This should ensure the item the order has stored locally has the meta change without a reload.

**Comment by sideburn** [comment]
I tried your suggestion. But first off I could not find a way to make woocommerce_order_status_completed pass two args. it only passes the order ID.

But with that said I tried the code below and I get the same issue. I see "hello world" appear on the order details webpage but not in the notification emails.

Code below:

 add_action('woocommerce_order_status_completed', array($this, 'action_woocommerce_order_status_completed'));		
 public function action_woocommerce_order_status_completed($order_id) {
	$order = new WC_Order($order_id);			
	$items = $order->get_items();
	foreach ($items as $key => $value) {
		$value->add_meta_data( $key,  "hello world");
		$value->save();
	}
}
